### PR TITLE
object_manager: Implement Chunk Batching on Push

### PR DESCRIPTION
**Pull request by barakmich**
## Why are these changes needed?

Currently, the object manager queues up every chunk (and copies its data into the appropriate outgoing message) whenever a push happens. For large objects, this can use a lot of memory and can block the io_service with lots of outgoing requests.

This change sets a maximum number of chunks to send at once, at which time the Push will wait for a response from the last chunk and then kick off the next batch.

## Related issue number

Related to #9432 

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -193,6 +193,11 @@ RAY_CONFIG(int, object_manager_repeated_push_delay_ms, 60000)
 /// chunks exceeds the number of available sending threads.
 RAY_CONFIG(uint64_t, object_manager_default_chunk_size, 1000000)
 
+/// The number of chunks to send at once before waiting for confirmation.
+/// More chunks means less latency, fewer chunks means better multiplexing of
+/// RPC calls.
+RAY_CONFIG(uint64_t, object_manager_default_batch_size, 100)
```
Comment (ericl) [inline src/ray/common/ray_config_def.h:199]: Looks like a good start, would be interesting to verify if this does indeed fix the broadcast issue. I think we'd probably want a sliding-window (not just batches) for more efficiency in a long term solution though.

Comment (ericl) [reply src/ray/common/ray_config_def.h:199]: There's also the question of how to tune this parameter, maybe we just run some experiments and see what a reasonable upper bound is (probably it's not more than a few GB is my guess).

Comment (stephanie-wang) [reply src/ray/common/ray_config_def.h:199]: Hmm this limit also only applies per object and sender-receiver pair, right? So we could still have a lot more chunks in flight for a given sender depending on what requests it gets. I think we should clarify this in the description.

File: src/ray/object_manager/object_manager.cc [hunk 2]
```diff
@@ -460,28 +460,58 @@ void ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id) {
...
+                                   const ClientID &client_id, uint64_t data_size,
+                                   uint64_t metadata_size, uint64_t from_chunk_index,
+                                   uint64_t num_chunks,
+                                   std::shared_ptr<rpc::ObjectManagerClient> rpc_client) {
+  uint64_t max_batch = RayConfig::instance().object_manager_default_batch_size();
+  uint64_t n = 1;
+
+  uint64_t index = from_chunk_index;
+
+  while (true) {
```
Comment (stephanie-wang) [inline src/ray/object_manager/object_manager.cc:483]: I found the logic in this loop a bit hard to follow. Could you add some comments?

File: src/ray/object_manager/object_manager.cc [hunk 3]
```diff
@@ -460,28 +460,58 @@ void ObjectManager::Push(const ObjectID &object_id, const ClientID &client_id) {
...
+    if (n > max_batch) {
+      break;
+    }
+    if (index >= num_chunks) {
+      break;
+    }
+    uint64_t send_num_chunks = 0;
```
Comment (stephanie-wang) [inline src/ray/object_manager/object_manager.cc:490]: continuation_num_chunks (to match the other function) seems like a better name here.

**Comment by ericl** [comment]
Closing stale PR. Please re-open if you still want to merge it.
